tempesta_fw/http.c

```diff
@@ -383,18 +369,636 @@ tfw_http_send_504(TfwHttpReq *req, const char *reason)
+{
+	TFW_DBG2("%s: conn=[%p]\n", __func__, srv_conn);
+
+	if (tfw_http_conn_need_fwd(srv_conn))
+		__tfw_http_conn_fwd_unsent(srv_conn, equeue);
+}
+
+/*
+ * Forward the request @req to server connection @srv_conn.
+ *
+ * The request is added to the server connection's forwarding queue.
+ * If forwarding is on hold at the moment, then the request will be
+ * forwarded later. Otherwise, forward the request to the server now.
+ *
+ * Forwarding to a server is considered to be on hold after
+ * a non-idempotent request is forwarded. The hold is removed when
+ * a response is received to the holding request. The hold is also
+ * removed when the holding non-idempotent request is followed by
+ * another request from the same client. Effectively, that re-enables
+ * pipelining. See RFC 7230 6.3.2.
+ *
+ * Requests must be forwarded in the same order they are put in the
+ * queue, and so it must be done under the queue lock, otherwise
+ * pairing of requests with responses may get broken. Take a simple
+ * scenario. CPU-1 locks the queue, adds a request to it, unlocks
+ * the queue. CPU-2 does the same after CPU-1 (the queue was locked).
+ * After that CPU-1 and CPU-2 are fully concurrent. If CPU-2 happens
+ * to proceed first with forwarding, then pairing gets broken.
+ */
+static void
+tfw_http_req_fwd(TfwSrvConn *srv_conn, TfwHttpReq *req)
+{
+	LIST_HEAD(equeue);
+
+	TFW_DBG2("%s: srv_conn=[%p], req=[%p]\n", __func__, srv_conn, req);
+	BUG_ON(!(TFW_CONN_TYPE(srv_conn) & Conn_Srv));
+
+	spin_lock(&srv_conn->fwd_qlock);
+	list_add_tail(&req->fwd_list, &srv_conn->fwd_queue);
+	srv_conn->qsize++;
+	if (tfw_http_req_is_nip(req))
+		__tfw_http_req_nip_enlist(srv_conn, req);
+	if (tfw_http_conn_on_hold(srv_conn)) {
+		spin_unlock(&srv_conn->fwd_qlock);
+		return;
+	}
+	__tfw_http_conn_fwd_unsent(srv_conn, &equeue);
+	spin_unlock(&srv_conn->fwd_qlock);
+
+	if (!list_empty(&equeue))
+		tfw_http_req_zap_error(&equeue);
+}
+
+/*
+ * Treat a possible non-idempotent request in case of a connection
+ * repair (re-send or re-schedule).
+ *
+ * A non-idempotent request that was forwarded but not responded to
+ * is not re-sent or re-scheduled by default. Configuration option
+ * can be used to have that request re-sent or re-scheduled.
+ *
+ * As forwarding is paused after a non-idempotent request is sent,
+ * there can be only one such request among forwarded requests, and
+ * that's @srv_conn->msg_sent.
+ *
+ * Note: @srv_conn->msg_sent may change in result.
+ */
+static inline void
+tfw_http_conn_treatnip(TfwSrvConn *srv_conn, struct list_head *equeue)
+{
+	TfwServer *srv = (TfwServer *)srv_conn->peer;
+	TfwHttpReq *req_sent = (TfwHttpReq *)srv_conn->msg_sent;
+
+	if (req_sent && tfw_http_req_is_nip(req_sent)
+	    && likely(!(srv->sg->flags & TFW_SRV_RETRY_NIP)))
+	{
+		BUG_ON(list_empty(&req_sent->nip_list));
+		srv_conn->msg_sent =
+			/* list_is_first(&req_sent->fwd_list, fwd_queue); */
```

Comment:
Maybe just say that this is implementation of `list_is_first()`?.. Now it looks like commented out line of code